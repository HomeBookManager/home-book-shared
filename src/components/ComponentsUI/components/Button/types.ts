// types
import { TIconProps } from '../Icon/Icon';

export type TButtonIcon =
  | (Pick<TIconProps, 'iconComponent'> & {
      applyFill?: boolean;
      applyStroke?: boolean;
    })
  | null;

export type TButtonIconArgs = {
  placement: 'start' | 'end';
  src: TButtonIcon;
};
