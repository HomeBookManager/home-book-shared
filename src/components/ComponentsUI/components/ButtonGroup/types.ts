// types
import { TButtonProps } from '../Button/Button';

export type TButtonGroupPickedButtonProps = Pick<
  TButtonProps,
  | 'className'
  | 'color'
  | 'disabled'
  | 'disabledRippleEffect'
  | 'e2eAttribute'
  | 'e2eValue'
  | 'forcedHover'
  | 'size'
  | 'style'
  | 'type'
  | 'variant'
>;

export type TButtonGroupContext = TButtonGroupPickedButtonProps | undefined;
