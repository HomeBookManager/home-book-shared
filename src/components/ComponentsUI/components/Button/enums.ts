export enum ButtonColor {
  primary = 'primary',
  secondary = 'secondary',
  success = 'success',
  warning = 'warning',
  error = 'error',
}

export enum ButtonSize {
  small = 'small',
  medium = 'medium',
  large = 'large',
}

export enum ButtonVariant {
  text = 'text',
  contained = 'contained',
  outlined = 'outlined',
}
