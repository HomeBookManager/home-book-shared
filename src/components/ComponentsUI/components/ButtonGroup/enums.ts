export enum ButtonGroupOrientation {
  horizontal = 'horizontal',
  vertical = 'vertical',
}
