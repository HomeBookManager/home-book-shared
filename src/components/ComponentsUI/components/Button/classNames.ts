import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from './enums';

export const className = 'Button';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      forcedHover: `${className}--forced-hover`,
      fullwidth: `${className}--full-width`,
      ...mapValues(ButtonColor, (color) => `${className}--${color}`),
      ...mapValues(ButtonSize, (size) => `${className}--${size}`),
      ...mapValues(ButtonVariant, (variant) => `${className}--${variant}`),
    },
  },
  icon: {
    name: `${className}__icon`,
    modificators: {
      applyFill: `${className}__icon--apply-fill`,
      applyStroke: `${className}__icon--apply-stroke`,
      end: `${className}__icon--end`,
      start: `${className}__icon--start`,
      ...mapValues(ButtonSize, (size) => `${className}__icon--${size}`),
    },
  },
};
