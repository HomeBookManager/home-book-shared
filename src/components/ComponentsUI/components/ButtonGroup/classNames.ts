import { mapValues } from 'lodash';

// types
import { ButtonColor, ButtonVariant } from '../Button/enums';
import { ButtonGroupOrientation } from './enums';

export const className = 'ButtonGroup';

export const classNames = {
  [className]: {
    name: className,
    modificators: {
      ...mapValues(
        ButtonGroupOrientation,
        (orientation) => `${className}--${orientation}`,
      ),
    },
  },
  button: {
    name: `${className}__button`,
    modificators: {
      forcedHover: `${className}__button--forced-hover`,
      ...mapValues(ButtonColor, (color) => `${className}__button--${color}`),
      ...mapValues(
        ButtonGroupOrientation,
        (orientation) => `${className}__button--${orientation}`,
      ),
      ...mapValues(
        ButtonVariant,
        (variant) => `${className}__button--${variant}`,
      ),
    },
  },
};
