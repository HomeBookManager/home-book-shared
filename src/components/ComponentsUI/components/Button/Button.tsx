import cx from 'classnames';
import { ButtonHTMLAttributes, ReactNode, Ref, forwardRef } from 'react';

// components
import E2EDataAttribute from '../../../E2EDataAttributes/E2EDataAttribute';

// hooks
import { useButtonGroupContext } from './hooks/useButtonGroupContext';
import { useClickInteraction } from './hooks/useClickInteraction';
import { useIcon } from './hooks/useIcon';
import { useRippleEffect } from '../../../../hooks/useRippleEffect/useRippleEffect';

// others
import { className as classNameButton, classNames } from './classNames';

// styles
import './styles/button.scss';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from './enums';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import { TButtonIcon } from './types';

export type TButtonProps = Omit<
  ButtonHTMLAttributes<HTMLElement>,
  'disabled'
> & {
  children?: ReactNode;
  className?: string;
  color?: ButtonColor;
  disabled?: boolean;
  disabledRippleEffect?: boolean;
  e2eAttribute?: string;
  e2eValue?: number | string;
  endIcon?: TButtonIcon;
  forcedHover?: boolean;
  fullWidth?: boolean;
  onClick?: () => void;
  ref?: Ref<HTMLButtonElement>;
  size?: ButtonSize;
  startIcon?: TButtonIcon;
  variant?: ButtonVariant;
};

export const Button = forwardRef<HTMLButtonElement, TButtonProps>(
  (
    {
      children,
      className = '',
      color = ButtonColor.primary,
      disabled = false,
      disabledRippleEffect = false,
      endIcon = null,
      e2eAttribute = E2EAttribute.button,
      e2eValue = '',
      forcedHover = false,
      fullWidth = false,
      onClick,
      size = ButtonSize.medium,
      startIcon = null,
      type = 'button',
      variant = ButtonVariant.contained,
      ...restProps
    },
    ref,
  ) => {
    const {
      classNameButtonGroup,
      colorButtonGroup,
      disabledRippleEffectButtonGroup,
      forcedHoverButtonGroup,
      sizeButtonGroup,
      variantButtonGroup,
      ...restButtonGroupProps
    } = useButtonGroupContext();

    const Icon = useIcon(size);

    const { rippleEffect, triggerRippleEffect } = useRippleEffect(
      classNames[classNameButton].name,
    );

    const onClickHandler = useClickInteraction(
      disabledRippleEffect || disabledRippleEffectButtonGroup,
      onClick,
      triggerRippleEffect,
    );

    return (
      <E2EDataAttribute type={e2eAttribute} value={e2eValue}>
        <button
          className={cx(
            className,
            classNameButtonGroup,
            classNames[classNameButton].name,
            { [classNames[classNameButton].modificators.fullwidth]: fullWidth },
            {
              [classNames[classNameButton].modificators.forcedHover]:
                forcedHover || forcedHoverButtonGroup,
            },
            classNames[classNameButton].modificators[colorButtonGroup || color],
            classNames[classNameButton].modificators[sizeButtonGroup || size],
            classNames[classNameButton].modificators[
              variantButtonGroup || variant
            ],
          )}
          disabled={disabled}
          onClick={onClickHandler}
          ref={ref}
          type={type}
          {...restButtonGroupProps}
          {...restProps}
        >
          {startIcon && <Icon placement="start" src={startIcon} />}
          {children}
          {endIcon && <Icon placement="end" src={endIcon} />}
          {rippleEffect}
        </button>
      </E2EDataAttribute>
    );
  },
);

export default Button;
