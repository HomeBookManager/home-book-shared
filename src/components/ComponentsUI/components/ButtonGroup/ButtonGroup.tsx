import cx from 'classnames';
import {
  HTMLAttributes,
  ReactElement,
  Ref,
  createContext,
  forwardRef,
} from 'react';

// components
import E2EDataAttribute from '../../../E2EDataAttributes/E2EDataAttribute';

// others
import { className as classNameButtonGroup, classNames } from './classNames';

// styles
import './styles/button-group.scss';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from '../Button/enums';
import { ButtonGroupOrientation } from './enums';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import { TButtonGroupContext, TButtonGroupPickedButtonProps } from './types';
import { TButtonProps } from '../Button/Button';

// utils
import { composeButtonClassNames } from './utils';

export const ButtonGroupContext = createContext<TButtonGroupContext>(undefined);

export type TButtonGroupProps = HTMLAttributes<HTMLDivElement> &
  TButtonGroupPickedButtonProps & {
    children: ReactElement<TButtonProps> | Array<ReactElement<TButtonProps>>;
    className?: string;
    classNameButton?: string;
    orientation?: ButtonGroupOrientation;
    ref?: Ref<HTMLDivElement>;
  };

export const ButtonGroup = forwardRef<HTMLDivElement, TButtonGroupProps>(
  (
    {
      children,
      className = '',
      classNameButton = '',
      color = ButtonColor.primary,
      disabled = false,
      disabledRippleEffect = false,
      e2eAttribute = E2EAttribute.buttonGroup,
      e2eValue = '',
      forcedHover = false,
      orientation = ButtonGroupOrientation.horizontal,
      size = ButtonSize.medium,
      style = {},
      type = 'button',
      variant = ButtonVariant.contained,
      ...restProps
    },
    ref,
  ) => {
    const buttonClassNames = composeButtonClassNames(
      classNameButton,
      color,
      forcedHover,
      orientation,
      variant,
    );

    return (
      <E2EDataAttribute type={e2eAttribute} value={e2eValue}>
        <div
          className={cx(
            className,
            classNames[classNameButtonGroup].name,
            classNames[classNameButtonGroup].modificators[orientation],
          )}
          ref={ref}
          style={style}
          {...restProps}
        >
          <ButtonGroupContext.Provider
            value={{
              className: buttonClassNames,
              color,
              disabled,
              disabledRippleEffect,
              forcedHover,
              size,
              type,
              variant,
            }}
          >
            {children}
          </ButtonGroupContext.Provider>
        </div>
      </E2EDataAttribute>
    );
  },
);

export default ButtonGroup;
