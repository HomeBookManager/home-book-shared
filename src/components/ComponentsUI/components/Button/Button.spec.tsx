import { fireEvent, render, waitFor } from '@testing-library/react';

// components
import Button from './Button';
import { IconUpOutlined } from '../Icon/components';

// core
import { ButtonGroupContext } from '../ButtonGroup/ButtonGroup';

// others
import { className as classNameButton, classNames } from './classNames';
import { RIPPLE_EFFECT_MODIFICATOR } from '../../../../hooks/useRippleEffect/constants';

// types
import { ButtonColor, ButtonSize, ButtonVariant } from './enums';
import { E2EAttribute } from '../../../E2EDataAttributes/enums';
import { IconShape } from '../Icon/enums';
import { TButtonGroupContext } from '../ButtonGroup/types';

// utils
import { getByE2EAttribute } from '../../../../tests/testHelpers';
import { getDataTestAttribute } from '../../../E2EDataAttributes/utils';
import { enumToArray } from '../../../../utils/transform/enumToArray';

const className = 'className';
const classNameButtonGroup = 'classNameButtonGroup';
const content = 'Click';
const mockCallBack = jest.fn();

describe('Button behaviors', () => {
  it('should render rippleEffect after click', async () => {
    // before
    const { container } = render(<Button>{content}</Button>);

    // find
    const button = getByE2EAttribute(container, E2EAttribute.button);

    // action
    fireEvent.click(button);

    // result
    await waitFor(
      () => {
        expect(button.lastChild).toHaveClass(
          `${classNames[classNameButton].name}--${RIPPLE_EFFECT_MODIFICATOR}`,
        );
      },
      { timeout: 100 },
    );
  });

  it('should provide data from button group contexxt', async () => {
    // mock
    const value = {
      className: classNameButtonGroup,
      color: ButtonColor.secondary,
      disabled: true,
      disabledRippleEffect: true,
      forcedHover: true,
      size: ButtonSize.large,
      type: 'submit',
      variant: ButtonVariant.text,
    } as TButtonGroupContext;

    // before
    const { container } = render(
      <ButtonGroupContext.Provider value={value}>
        <Button>{content}</Button>
      </ButtonGroupContext.Provider>,
    );

    // find
    const button = getByE2EAttribute(container, E2EAttribute.button);

    // action
    fireEvent.click(button);

    // result
    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveClass(
      classNameButtonGroup,
    );

    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveClass(
      classNames[classNameButton].modificators[ButtonColor.secondary],
    );

    expect(getByE2EAttribute(container, E2EAttribute.button)).toBeDisabled();

    await waitFor(
      () => {
        expect(button.lastChild).toHaveTextContent(content);
      },
      { timeout: 100 },
    );

    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveClass(
      classNames[classNameButton].modificators.forcedHover,
    );

    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveClass(
      classNames[classNameButton].modificators[ButtonSize.large],
    );

    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveAttribute(
      'type',
      'submit',
    );

    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveClass(
      classNames[classNameButton].modificators[ButtonVariant.text],
    );
  });
});

describe('Button props', () => {
  it('should pass children', () => {
    // before
    const { container } = render(<Button>{content}</Button>);

    // result
    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveTextContent(
      content,
    );
  });

  it('should pass className', () => {
    // before
    const { container } = render(
      <Button className={className}>{content}</Button>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveClass(
      className,
    );
  });

  it('should pass color', () => {
    // mock
    const colors = enumToArray(ButtonColor) as Array<ButtonColor>;

    // before
    const { container } = render(
      <>
        {colors.map((color) => (
          <Button color={color} e2eValue={color} key={color}>
            {content}
          </Button>
        ))}
      </>,
    );

    // result
    colors.forEach((color) =>
      expect(
        getByE2EAttribute(container, E2EAttribute.button, color),
      ).toHaveClass(classNames[classNameButton].modificators[color]),
    );
  });

  it('should pass disableRippleEffect', async () => {
    // before
    const { container } = render(
      <Button disabledRippleEffect>{content}</Button>,
    );

    // find
    const button = getByE2EAttribute(container, E2EAttribute.button);

    // action
    fireEvent.click(button);

    // result
    await waitFor(
      () => {
        expect(button.lastChild).toHaveTextContent(content);
      },
      { timeout: 100 },
    );
  });

  it('should pass e2eAttribute', () => {
    // before
    const { container } = render(
      <Button e2eAttribute={E2EAttribute.button}>{content}</Button>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.button),
    );
  });

  it('should pass e2eValue', () => {
    // before
    const { container } = render(
      <Button e2eValue="e2eValue">{content}</Button>,
    );

    // result
    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveAttribute(
      getDataTestAttribute(E2EAttribute.button),
      'e2eValue',
    );
  });

  it('should pass endIcon', () => {
    // mock
    const endIcon = {
      applyFill: true,
      iconComponent: IconUpOutlined,
      iconShape: IconShape.upOutlined,
    };

    // before
    const { container } = render(<Button endIcon={endIcon}>{content}</Button>);

    // result
    expect(
      getByE2EAttribute(container, E2EAttribute.button).lastChild,
    ).toHaveClass(classNames.icon.name);
  });

  it('should pass forcedHover', () => {
    // before
    const { container } = render(<Button forcedHover>{content}</Button>);

    // result
    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveClass(
      classNames[classNameButton].modificators.forcedHover,
    );
  });

  it('should pass fullWidth', () => {
    // before
    const { container } = render(<Button fullWidth>{content}</Button>);

    // result
    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveClass(
      classNames[classNameButton].modificators.fullwidth,
    );
  });

  it('should pass onClick', () => {
    // before
    const { container } = render(
      <Button onClick={mockCallBack}>{content}</Button>,
    );

    // action
    fireEvent.click(getByE2EAttribute(container, E2EAttribute.button));

    // result
    expect(mockCallBack.mock.calls.length).toBe(1);
  });

  it('should pass size', () => {
    // mock
    const sizes = enumToArray(ButtonSize) as Array<ButtonSize>;

    // before
    const { container } = render(
      <>
        {sizes.map((size) => (
          <Button e2eValue={size} key={size} size={size}>
            {content}
          </Button>
        ))}
      </>,
    );

    // result
    sizes.forEach((size) =>
      expect(
        getByE2EAttribute(container, E2EAttribute.button, size),
      ).toHaveClass(classNames[classNameButton].modificators[size]),
    );
  });

  it('should pass startIcon', () => {
    // mock
    const startIcon = {
      applyFill: true,
      iconComponent: IconUpOutlined,
      iconShape: IconShape.upOutlined,
    };

    // before
    const { container } = render(
      <Button startIcon={startIcon}>{content}</Button>,
    );

    // result
    expect(
      getByE2EAttribute(container, E2EAttribute.button).firstChild,
    ).toHaveClass(classNames.icon.name);
  });

  it('should pass type', () => {
    // before
    const { container } = render(<Button type="submit">{content}</Button>);

    // result
    expect(getByE2EAttribute(container, E2EAttribute.button)).toHaveAttribute(
      'type',
      'submit',
    );
  });

  it('should pass variant', () => {
    // mock
    const variants = enumToArray(ButtonVariant) as Array<ButtonVariant>;

    // before
    const { container } = render(
      <>
        {variants.map((variant) => (
          <Button e2eValue={variant} key={variant} variant={variant}>
            {content}
          </Button>
        ))}
      </>,
    );

    // result
    variants.forEach((variant) =>
      expect(
        getByE2EAttribute(container, E2EAttribute.button, variant),
      ).toHaveClass(classNames[classNameButton].modificators[variant]),
    );
  });
});

describe('Button behaviors', () => {
  it('should render Button', () => {
    // before
    const { asFragment } = render(<Button>{content}</Button>);

    // result
    expect(asFragment()).toMatchSnapshot();
  });
});
