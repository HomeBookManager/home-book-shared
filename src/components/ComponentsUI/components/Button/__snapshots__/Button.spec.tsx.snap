// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button behaviors should render Button 1`] = `
<DocumentFragment>
  <button
    class="Button Button--primary Button--medium Button--contained"
    data-test-button=""
    type="button"
  >
    Click
  </button>
</DocumentFragment>
`;
